src/protocol/sort/secureapplyinv.rs

```diff
@@ -0,0 +1,124 @@
+use crate::{
+    error::BoxError,
+    ff::Field,
+    protocol::{
+        context::ProtocolContext,
+        reveal::reveal_a_permutation,
+        sort::ApplyInvStep::{RevealPermutation, ShuffleInputs, ShufflePermutation},
+    },
+    secret_sharing::Replicated,
+};
+use embed_doc_image::embed_doc_image;
+
+use super::{
+    apply::apply,
+    shuffle::{get_two_of_three_random_permutations, Shuffle},
+};
+use tokio::try_join;
+
+/// This is an implementation of ApplyInv (Algorithm 4) found in the paper:
+/// "An Efficient Secure Three-Party Sorting Protocol with an Honest Majority"
+/// by K. Chida, K. Hamada, D. Ikarashi, R. Kikuchi, N. Kiribuchi, and B. Pinkas
+/// https://eprint.iacr.org/2019/695.pdf
+/// This is a protocol that applies the inverse of a secret-shared permutation to a vector of secret-shared values
+/// Input: Each helpers know their own secret shares of input and permutation
+/// Output: At the end of the protocol, all helpers receive inputs after the permutation is applied
+#[derive(Debug)]
+#[embed_doc_image("secureapplyinv", "images/sort/secureapplyinv.png")]
+pub struct SecureApplyInv {}
+
+impl SecureApplyInv {
+    /// This algorithm applies a permutation to the `input` vector. The permutation is secret-shared,
+    /// and none of the helpers should learn it through this protocol.
+    /// To keep the permutation secret, it (and the inputs) are first randomly securely shuffled.
+    /// After this shuffle, the permutation can be revealed.
+    /// An adversary can only obtain a shuffled permutation, which is just a random permutation.
+    /// Steps
+    /// ![Secure Apply Inv steps][secureapplyinv]
+    /// 1. Generate random permutations using prss
+    /// 2. Secret shared permutation is shuffled with random permutations
+    /// 3. Secret shared value is shuffled using the same random permutations
+    /// 4. The permutation is revealed
+    /// 5. All helpers call `apply` to apply the permutation locally.
+    #[allow(dead_code)]
+    pub async fn execute<F: Field>(
+        ctx: &ProtocolContext<'_, F>,
+        input: &'_ mut Vec<Replicated<F>>,
+        sort_permutation: &'_ mut Vec<Replicated<F>>,
+    ) -> Result<(), BoxError> {
+        let mut random_permutations =
+            get_two_of_three_random_permutations(input.len(), &ctx.prss());
+        let mut random_permutations_copy = random_permutations.clone();
+
+        let mut shuffle_inputs = Shuffle::new(input);
+        let mut shuffle_permutation = Shuffle::new(sort_permutation);
+
+        let (_, _) = try_join!(
```

Comment:
You don't need to use the macro version. You can just use `try_join` from the futures crate:
```
use futures::future::try_join; 
```